hphp/runtime/vm/jit/vasm-lower-inl.h

```diff
@@ -38,14 +44,23 @@ void vasm_lower(Vunit& unit, Vlower lower_impl) {
   // iterators.
   auto& blocks = unit.blocks;
 
+  VLS env{unit, true};
+  auto vregConstraints = detail::computeDominatingConstraints(unit);
+
   // The vlower() implementations, and `lower', may allocate scratch blocks and
   // modify instruction streams, so we cannot use standard iterators here.
   PostorderWalker{unit}.dfs([&] (Vlabel b) {
     assertx(!blocks[b].code.empty());
 
+    env.allow_vreg = vregConstraints[b] != VregConstraint::Restrict;
```

Comment:
Note: I implemented this as suggested, but it drops the nesting concept a little. I use the integer to indicate the nested depth of the `vregrestrict`s when calculating the `vregConstraints`. But then when using this `env.allow_vreg` bool, we lose the nest depth. In a scenario where `vregConstraints[b] == 3`, a `vregallow` within a block will allow new vregs.

Is that you intended behavior? I think it needs to be.